test/raft/replication.hh

```diff
@@ -834,7 +842,7 @@ future<> raft_cluster<Clock>::stop_server(size_t id) {
     cancel_ticker(id);
     co_await _servers[id].server->abort();
     if (_snapshots->contains(to_raft_id(id))) {
-        BOOST_CHECK((*_snapshots)[to_raft_id(id)].size() <= 2);
+        BOOST_CHECK_LE((*_snapshots)[to_raft_id(id)].size(), _max_snapshots);
```

Comment:
ok

Reply:
is it expected to have many snapshots with concurrent add_entries?

Reply:
I do not see why it would. We apply and create snapshots sequentially, so at any time there should be only one snapshot installed. Another one may come from snapshot transfer.

Reply:
Is there something preventing `_output.snp `in `fsm::apply_snapshot` to be overwritten?

* we approach memory limit, so `applier_fiber` runs snapshots on each entry
* it gets scheduled two times one after the other, applying two commands and taking snapshots for each of them
* when `io_fiber` eventually runs it sees only the last snapshot in `batch.snp`, so the previous one is 'leaked'

Reply:
Well, then you found a leak that should be fixed. Please write a test that triggers it on top of the master and then we can think about the fix.

Reply:
I've just added this on master
```
diff --git a/raft/fsm.cc b/raft/fsm.cc
index da18f1a963..478cf10bc5 100644
--- a/raft/fsm.cc
+++ b/raft/fsm.cc
@@ -997,6 +997,9 @@ bool fsm::apply_snapshot(snapshot_descriptor snp, size_t trailing, bool local) {
     // If the snapshot is local, _commit_idx is larger than snp.idx.
     // Otherwise snp.idx becomes the new commit index.
     _commit_idx = std::max(_commit_idx, snp.idx);
+    if (_output.snp.has_value()) {
+        on_internal_error(logger, "leaked snapshot");
+    }
     _output.snp.emplace(fsm_output::applied_snapshot{snp, local, current_snp.id});
     size_t units = _log.apply_snapshot(std::move(snp), trailing);
     if (is_leader()) {
```
and `basic_generator_test` failed after several retries. Is it sufficient?

Reply:
This is on master? If yes I think it is.

Reply:
OTOH the fix I was thinking about would change the code here (return not yet applied snapshot id for removing), so `on_internal_error` will be dropped and we will have to invent another test.